### ISSUE TITLE
Wrong nodes can be returned by NodeDataRepository::reduceNodeVariantsByWorkspacesAndDimensions

### ISSUE THREAD

**Issue by gjwnc** [neos/neos-development-collection#2664]
### Description

The method https://github.com/neos/neos-development-collection/blob/2516e40258bc9f21fe8a466bf867ab0c7682e7a8/Neos.ContentRepository/Classes/Domain/Repository/NodeDataRepository.php#L1288 might not return the correct nodes if the node dimension is not found in the method argument `$dimensions`.

### Steps to Reproduce

For simplicity the arguments for `reduceNodeVariantsByWorkspacesAndDimensions` must be like the following:  
* `$nodes` An array of two nodes with the same identifier in the live workspace. The dimensions are `{"language":"de"}` for the first node and `{"language":"en"}` for the second. The order of the nodes is relevant.
* `$workspaces` An array containing only the live workspace.
* `$dimensions` An array of the form `{"language": "en"}`

#### Expected behavior

In the final returned array `$reducedNodes` only the second node should be returned, since this node matches the requested dimensions from `$dimensions`.

#### Actual behavior

The first node is returned.

#### Problem analysis

In the first iteration of the `$nodes` loop the `de` node will be iterated. Since both, the node dimensions and `$dimensions` contain `language` the inner loop is processed, see https://github.com/neos/neos-development-collection/blob/2516e40258bc9f21fe8a466bf867ab0c7682e7a8/Neos.ContentRepository/Classes/Domain/Repository/NodeDataRepository.php#L1320.  
The `array_search` will fail, because `de` is not found in `$dimensions` and thus `$position` is false.
In line https://github.com/neos/neos-development-collection/blob/2516e40258bc9f21fe8a466bf867ab0c7682e7a8/Neos.ContentRepository/Classes/Domain/Repository/NodeDataRepository.php#L1334 the `$minimalDimensionPositionsByIdentifier` is empty since this is the first iteration, thus the condition is true and the node is added to `$reducedNodes`.

In the second iteration of the `$nodes` loop the `en` node is used. Again, the `language` dimension matches and thus `array_search` will find a valid position in `$dimension` which results in `$position` having the integer value `0`. Now the second condition on line 1334 is evaluated where `$dimensionPositions < $minimalDimensionPositionsByIdentifier[$identifier]` is false. So the node is ignored.
As an example, the comparision is done for the following two arrays:
```
$dimensionPositions = [
    "language": 0,
    "0": 1
];
$minimalDimensionPositionsByIdentifier[$identifier] = [
    "language": false,
    "0": 1
];
```

The very reason this fails, is because `0 < false` is false.

#### Possible solution

In case `array_search` doesn't find the dimension value, this node should be pushed back to the very end of the `$dimensionPositions[$dimensionName]` array, such that further `min()` calls will prefer matching nodes given a set of nodes and dimensionValues in a specific order.

The solution could be to change
```
$position = array_search($nodeDimensionValue, $dimensionValues);
```
to
```
$position = array_search($nodeDimensionValue, $dimensionValues);
if($position === false){
    $position = PHP_INT_MAX;
}
```

#### Sidenote optimization

The call to `array_map` for the workspaces is run for every node, but does not depend on the current `$node`. So `array_map` can be moved before the `foreach ($nodes as $node) {` loop.

Neos: 4.3.3

Flow: 5.3.2

**Comment by gjwnc** [comment]
@kitsunet

**Comment by gjwnc** [comment]
Closing this, since this has already been merged by PR #2736.
